feat: production bundling (#2413)
vaadin/docs: articles/production/production-build.adoc

= Production Build

> The default production bundle can be used to not have to run front-end tools like npm and Vite if no add-ons or front-end files are used in the application.

Reviewer: ```suggestion
The pre-compiled production bundle can be used to not have to run front-end tools like npm and Vite if no add-ons or front-end files are used in the application.
```

> 

Reviewer: I propose to give here a link (from dev configuration chapter) or to list explicitly the reasons when Flow decides to make a new prod bundle.

The bundle can be optimized to use <<{articles}/advanced/lazy-loading#,lazy loading>> of components by building an optimized bundle.

To create an optimized production build, run the following command:

[source,terminal]
----
mvn clean package -Pproduction -Dforce.production.build=true
----

This builds a `JAR` or `WAR` file with all the dependencies and bundled front-end resources, ready to be deployed.
The file can be found in the `target` folder after the build completes.

> To generate a production build by using a pre-created production bundle that is applicable for the application run:

Reviewer: ```suggestion
To generate a production build by using a pre-compiled production bundle that is applicable for the application run:
```

= Production Build

The pre-compiled production bundle can be used to not have to run front-end tools like npm and Vite if no add-ons or front-end files are used in the application.

Vaadin generates an application bundle in the following situations:

- An npm/pnpm package is added with `@NpmPackage` or directly into [filename]`package.json`;
- CSS or JavaScript is added with `@CssImport`, `@JsModule` or `@JavaScript`;
- Vaadin add-on with front-end customizations is added;
- Custom theme packaged as a `JAR` dependency is added, if it defines any assets to be added to the project;

> - Exported web component is added.

Reviewer: ```suggestion
- Exported web component is added;
```

Author: Removed all `;` as they look strange on the ending of list items

- `@LoadDependenciesOnStartup` is defined for `AppShellConfigurator`

The bundle can be optimized to use <<#_lazy_loading_javascript,lazy loading>> of components by forcing a build of the production bundle.

To force creation of an optimized production bundle, run the following command:

[source,terminal]
----

> mvn clean package -Pproduction -Dforce.production.build=true

Reviewer: Suggestion from DX tester: mention `forceProductionBuild` also here and give an example how to set it with the `vaadin-maven-plugin`.

Author: added

> mvn clean package -Pproduction -Dforce.production.build=true

Reviewer: This is might be a separate PR, but this parameter listed below in the mvn example snipped
```
        <properties>
            <vaadin.productionMode>true</vaadin.productionMode>
        </properties>
```
has no effect anymore, so should be removed.

Author: removed

Reviewer 2: I think it's always better to state the simple use case first, and then suggest the more advanced use case after that. Here, it would be better to state the simple case first without `forceProductionBuild`, and then clarify to advanced users how they can accomplish the more complex thing.

=== dance
This goal is synonymous with the `clean-frontend` goal.

> === Lazy Loading JavaScript

Reviewer: I propose to reorganise it a bit:
let this be with indentation `==` and title "Bundle Chunks Loading Optimisations" with two sub chapters "Lazy Loading (On Demand)" and "Eager Loading"

Author: Went with the higher level `Bundle Component Loading Optimisations` as one doesn't actually touch chunks.

> A production build scans for `Routes` and make lazy loaded bundles of the components used in the routes.

Reviewer: ```suggestion
A production build scans for `Routes` and loads lazily the components (included into bundle's chunks) used in the routes.
```

Author: omitted talk about chunks as that is not important and makes it more confusing than helpful

> A production build scans for `Routes` and make lazy loaded bundles of the components used in the routes.

Reviewer: ```suggestion
A production build scans for `Routes` and loads lazily the components (included into bundle's chunks) used in the routes.
```

By default, only the routes `""` and `"login"` are eager and every other route components become lazy loaded.

> With the default production bundle all components are loaded eagerly apart from the heavy components `Map`, `Charts`, `Spreadsheet` and `RichTextEditor`.

Reviewer: ```suggestion
With the pre-compiled production bundle all components are loaded eagerly apart from the heavy components `Map`, `Charts`, `Spreadsheet` and `RichTextEditor`.
```
